weather_dl/download_pipeline/fetcher.py

```diff
@@ -44,29 +44,122 @@ class Fetcher(beam.DoFn):
         client_name: The name of the download client to construct per each request.
         manifest: A manifest to keep track of the status of requests
         store: To manage where downloads are persisted.
+        optimise_download: Optimised the download.
     """
 
     client_name: str
     manifest: Manifest = NoOpManifest(Location('noop://in-memory'))
     store: t.Optional[Store] = None
+    optimise_download: bool = False
 
     def __post_init__(self):
         if self.store is None:
             self.store = FSStore()
 
+    def expand(self, element):
+        request = (
+            element
+            | 'EachPartition' >> beam.ParDo(EachPartition())
+        )
+
+        if self.optimise_download:
+            downloaded_data = (
+                request
+                | 'Fetch' >> beam.ParDo(FetchData(self.client_name, self.manifest, self.store))
+                | 'Download' >> beam.ParDo(DownloadData(self.client_name, self.manifest, self.store))
+            )
+        else:
+            downloaded_data = (
+                request
+                | 'Fetch+Download' >> beam.ParDo(Retrieve(self.client_name, self.manifest, self.store))
+            )
+        (
+            downloaded_data
+            | 'Upload' >> beam.ParDo(Upload(self.client_name, self.manifest, self.store))
+        )
+
+
+class EachPartition(beam.DoFn):
```

Comment:
For each of these DoFns, please add class docstrings (I assume you would anyway at a more advanced draft).

Reply:
Done.